### ISSUE TITLE
Support for SMB as FS-Provider

### ISSUE THREAD

**Issue by helmut72** [drakkan/sftpgo#148]
Please add support for SMB as a FS-Provider. SFTPGo becomes a fast (Golang!) Middleware to serve SMB Shares with secure and suitable protocols (SFTP, WebDAV, FTPS) for transfers over the Internet.

SFTPGo should use the user credentials to allow mapping to the SMB FS-Provider. It should work like this:
https://docs.nextcloud.com/server/latest/admin_manual/configuration_files/external_storage_configuration_gui.html

Of course, this will only work, if the authentication to SFTPGo use username and password. Therefore, SMB FS-Provider should be only enabled if at least authentication with username+password is used.

Maybe this becomes easier to implement, if SFTPGo just starts an external command (mount.cifs) with required arguments like username, password and mountpoint (with variables like username and groups) and send an unmount command after x seconds idle. Basically, a mounted SMB share is like a local file system share for SFTPGo. mount.cifs can provide all shares with the UID/GID that is used by SFTPGo.

Thank you!

**Comment by dkmbw** [comment]
I am REALLY thinking about forking SFTPGo. Although i do not like forking stuff, but atm everything gets stuffed into it and that creates security risks. Please think hard about including all that stuff (for me personally, webdav was already too much, but hey, i am just a user who really likes sftpgo.

On Top: CIFS is a REALLY COMPLICATED protocol which from my POV hs no place in public internet. there are OFTEN security issues with it.

**Comment by helmut72** [comment]
You misunderstand me. I've asked for CIFS as a storage **backend**. There are still 3 backends implemented: Local, S3 and Google.

Also I've asked just to run an external command, because a SMB mount is just a local filesystem for SFTPGo.

> which from my POV hs no place in public internet
Exactly. I'm 100% with you.

**Comment by drakkan** [comment]
> I am REALLY thinking about forking SFTPGo. Although i do not like forking stuff, but atm everything gets stuffed into it and that creates security risks. Please think hard about including all that stuff (for me personally, webdav was already too much, but hey, i am just a user who really likes sftpgo.
> 
> On Top: CIFS is a REALLY COMPLICATED protocol which from my POV hs no place in public internet. there are OFTEN security issues with it.

Hi,

most things can be disabled at build time, please take a look [here](https://github.com/drakkan/sftpgo/blob/master/docs/build-from-source.md).

For example a minimal build will result in a binary of about 13MB and it will not include the dependencies for the features you exclude. So you can customize the external libraries included inside the sftpgo binary

**Comment by drakkan** [comment]
> You misunderstand me. I've asked for CIFS as a storage **backend**. There are still 3 backends implemented: Local, S3 and Google.
> 

Hi, these backeds provide a multi platform API using a good supported Go library.

Using for example `mount.cifs` is platform specific, require an external system command etc..

My first answer is that you can do these kind of things using an hook. For example you can use a pre-login hook to check/mount a samba fs for a specific user.

I think that if there isn't a good Go library for SMB we cannot have SMB as fs provider. My suggestion is to use an hook. 

Anyway I'll leave this issue open for further discussions/ideas. Thank you 

> Also I've asked just to run an external command after authentication, because a SMB mount is a local filesystem for SFTPGo.
> 
> > which from my POV hs no place in public internet
> 
> Exactly. I'm 100% with you.

**Comment by helmut72** [comment]
> For example you can use a pre-login hook to check/mount a samba fs for a specific user.

I've overseen this. Thanks!

**Comment by drakkan** [comment]
Hi,

I did some searchs, there are several go libraries for samba but noone seems stable, mature and well supported for now.

The only viable approach seems to use `mount.cifs` on Linux. This cannot be included directly inside SFTPGo code for the reasons explained abobe, anyway you can use it via an hook.

You can write, for example, a bash script or an http server (if you have a lot of login) yourself, this could be an example flow for the hook:

1) it receives the username and the other user info as input 
2) it do a lookup to get the samba share and the path to mount 
3) it checks if the share is already mounted or wait if another mount is already in progress
4) it allows the login after the share is mounted

If you wrote this hook please send a pull request, it could be useful for others too and I will be glad to put it inside the [examples](https://github.com/drakkan/sftpgo/tree/master/examples). Thank you!

**Comment by helmut72** [comment]
Yes, will do it. My main purpose for mounting SMB is using it with WebDAV. SFTP is for local files only. Also the hook must receive the password, to fill the credentials for mount.cifs.

**Comment by tooptoop4** [comment]
https://github.com/hirochachacha/go-smb2 is good

**Comment by dynek** [comment]
@helmut72 did you ever end up writing a doc/example to mount a SMB share?

**Comment by helmut72** [comment]
No, for this usecase I continued with Apaches webdav and external auth module. But basically it's the same steps:

- read username + password
- check with ldapsearch if authentication would be successful
- check if mount exists, if not use username + password and mount with smbmount

In the meantime I no longer use it, because I want to leave my Synology in standby mode for longer and have added more storage space to my Nuc.

**Comment by tecosaur** [comment]
Some months ago I spent a while searching for a good enough SMB library in Go, but couldn't find one (maintained or not) that didn't involve incomplete functionality/severe compromises.

While far from mature enough to seriously consider, I've just stumbled across an interesting project that seems like it could finally provide a production-quality SMB server for Go: https://github.com/kardianos/gosmb.

Compared to options like `mount.cifs`, this allows for per-user virtual filesystems, and doesn't require running an external command.

The one downside seems to be cross-platform support, since it works by communicating with the [ksmbd](https://github.com/namjaejeon/ksmbd) Linux kernel module (without using cgo though!).
